### PR TITLE
Try to add ldap libraries to the docker images.

### PR DESCRIPTION
**Pull request by sandreenko**
I am trying to add a build time dependency between CoreFX library and OpenLdap.
In order to do so, I need OpenLdap library to exist in the docker images that we use for build.

I am not familiar with this infrastructure so please review as if a monkey wrote this change.

cc @janvorli 

link https://github.com/dotnet/runtime/pull/50540

### CODE REVIEW

File: eng/common/cross/build-android-rootfs.sh [hunk 1]
```diff
@@ -106,6 +106,7 @@ __AndroidPackages+=" libandroid-glob"
 __AndroidPackages+=" liblzma"
 __AndroidPackages+=" krb5"
 __AndroidPackages+=" openssl"
+# do I need ldap on android? how can I find android package name?
```
Comment (janvorli) [inline eng/common/cross/build-android-rootfs.sh:109]: The package name is `openldap`. You can find it by downloading the file below (e.g. http://termux.net/dists/stable/main/binary-aarch64/Packages for arm64) and then search for ldap.

File: eng/common/cross/build-rootfs.sh [hunk 2]
```diff
@@ -55,11 +55,14 @@ __UbuntuPackages+=" libcurl4-openssl-dev"
 __UbuntuPackages+=" libkrb5-dev"
 __UbuntuPackages+=" libssl-dev"
 __UbuntuPackages+=" zlib1g-dev"
+__UbuntuPackages+=" zlib1g-dev"
```
Comment (janvorli) [inline eng/common/cross/build-rootfs.sh:58]: This one was duplicated accidentally.

File: eng/common/cross/build-rootfs.sh [hunk 3]
```diff
@@ -55,11 +55,14 @@ __UbuntuPackages+=" libcurl4-openssl-dev"
...
+__UbuntuPackages+=" libldap2-dev"
 
 __AlpinePackages+=" curl-dev"
 __AlpinePackages+=" krb5-dev"
 __AlpinePackages+=" openssl-dev"
 __AlpinePackages+=" zlib-dev"
+__AlpinePackages+=" libldap2-dev"
```
Comment (janvorli) [inline eng/common/cross/build-rootfs.sh:65]: Alpine package is `openldap-dev`

File: eng/common/cross/build-rootfs.sh [hunk 4]
```diff
@@ -68,11 +71,13 @@ __FreeBSDPackages+=" icu"
 __FreeBSDPackages+=" libinotify"
 __FreeBSDPackages+=" lttng-ust"
 __FreeBSDPackages+=" krb5"
+__FreeBSDPackages+=" libslapi-2.4"
 
 __IllumosPackages="icu-64.2nb2"
 __IllumosPackages+=" mit-krb5-1.16.2nb4"
 __IllumosPackages+=" openssl-1.1.1e"
 __IllumosPackages+=" zlib-1.2.11"
+__IllumosPackages+=" ldap-0.5.11"
```
Comment (janvorli) [inline eng/common/cross/build-rootfs.sh:80]: @am11 can you please verify the Illumos package name?

Comment (sandreenko) [reply eng/common/cross/build-rootfs.sh:80]: found it here http://pkg.openindiana.org/hipster/en/search.shtml?token=ldap&action=Search

**Comment by sandreenko** [comment]
@janvorli could you please explain why we don't have ldap in [Interop libraries for Android ](https://github.com/dotnet/runtime/blob/350ee45ba750efad16b49c920092340e01eaa722/src/libraries/Common/src/Interop/Android/Interop.Libraries.cs )? Is it inherited from Unix or not used there? I guess it is not used then how do we run System.DirectoryServices.Protocols there? I guess it is excluded somehow but can't find information about it.

The question is if we want to add this dependency for android or not, the same goes for FreeBSD I think.

File: eng/common/cross/build-rootfs.sh (hunk 4 above)
Comment (am11) [reply eng/common/cross/build-rootfs.sh:79]: Thanks for the message @janvorli.

@sandreenko, we are currently using SmartOS' distro for the raw packages:
US: https://pkgsrc.joyent.co/packages/SmartOS/2020Q1/x86_64/All/
EU: https://pkgsrc.smartos.skylime.net/packages/SmartOS/2020Q1/x86_64/All/

the package which has libldap.so is http://pkgsrc.smartos.skylime.net/packages/SmartOS/2020Q1/x86_64/All/openldap-client-2.4.49.tgz. Assuming that's the lib we are interested in(?), I'd change it to:
```suggestion
__IllumosPackages+=" openldap-client-2.4.49"
```
btw, we can test it by running `eng/common/cross/build-rootfs.sh --use-mirror illumos` (or without mirror in US) from root of the repo on linux machine and then inspect/`find` stuff in `./.tools/rootfs/` directory.

```sh
# prereqs
$ sudo apt install g++ make libgmp3-dev libmpfr-dev libmpc-dev

# run the script for illumos
$ ./build-rootfs.sh --use-mirror illumos
# succeeds

$ find .tools/ -name *ldap.so
.tools/rootfs/x64/lib/krb5/plugins/kdb/kldap.so
.tools/rootfs/x64/lib/libkdb_ldap.so
.tools/rootfs/x64/lib/libldap.so
.tools/rootfs/x64/usr/lib/amd64/libldap.so
.tools/rootfs/x64/usr/lib/amd64/libsldap.so
.tools/rootfs/x64/usr/lib/libldap.so
.tools/rootfs/x64/usr/lib/libsldap.so
.tools/rootfs/x64/usr/lib/security/amd64/pam_ldap.so
.tools/rootfs/x64/usr/lib/security/pam_ldap.so
```

**Comment by am11** [comment]
> The question is if we want to add this dependency for android or not, the same goes for FreeBSD I think.

(I am not @janvorli but) IMO, we can add the dependencies as these are very early-stage setup script, and doesn't hurt to add an extra package which may or may not be consumed by the build. If we need them later on those platforms, it would be easier to consume them OOTB from the docker image (rather than going through this process: update this script, wait for `dotnet/dotnet-buildtools-prereqs-docker ` repo to create a new image, consume the image in dotnet/runtime).

**Comment by janvorli** [comment]
Android with CoreCLR is currently in a mode where we've made things work in the past, but we never test it or do anything with it.  Official Android support comes from the Mono in .NET. Looking at the history of the interop libraries for Android you've mentioned, I can see it was added for Mono purposes.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: eng/common/cross/build-rootfs.sh (hunk 4 above)
Comment (sandreenko) [reply eng/common/cross/build-rootfs.sh]: Thanks @am11, 

> btw, we can test it by running eng/common/cross/build-rootfs.sh --use-mirror illumos (or without mirror in US) from root of the repo on linux machine and then inspect/find stuff in ./.tools/rootfs/ directory.

Could you please clarify if we need to run something after "find .tools/ -name *ldap.so" to find where ".tools/rootfs/x64/usr/lib/amd64/libldap.so" comes from?

**Pull request merged**